[ARM][MVE] Invalid tail predication in LowOverheadLoop pass

statham-arm opened the pull request:
When a loop is converted into a low-overhead loop using tail predication via FPSCR.LTPSIZE, the MQPRCopy pseudo-instruction is expanded into either two VMOVD or a single MVE_VORR, depending on whether the values written to the lanes with a 'false' predicate matter. (MVE_VORR uses the ambient LTPSIZE predicate, so it won't write those lanes at all; the double VMOVD is slower but gets them right.)

This check was done based on whether the output of the MQPRCopy is live coming out of the loop. But it missed a case where the live-out value is not _itself_ an MQPRCopy, but is a predicated operation taking its false lanes from an MQPRCopy.

Fixes #162644, and adds a new MIR test case derived from the reproducer in that bug.

statham-arm committed:
[ARM][MVE] Invalid tail predication in LowOverheadLoop pass

When a loop is converted into a low-overhead loop using tail
predication via FPSCR.LTPSIZE, the MQPRCopy pseudo-instruction is
expanded into either two VMOVD or a single MVE_VORR, depending on
whether the values written to the lanes with a 'false' predicate
matter. (MVE_VORR uses the ambient LTPSIZE predicate, so it won't
write those lanes at all; the double VMOVD is slower but gets them
right.)

This check was done based on whether the output of the MQPRCopy is
live coming out of the loop. But it missed a case where the live-out
value is not _itself_ an MQPRCopy, but is a predicated operation
taking its false lanes from an MQPRCopy.

Fixes #162644, and adds a new MIR test case derived from the
reproducer in that bug.

statham-arm commented:
This is the second attempt at #163217. Instead of inhibiting tail predication when a problem is detected, it just adds the offending MQPRCopy to the existing list of ones that must be rewritten as two VMOVD, and keeps the rest of the loop tail-predicated as before.

If #163723 lands before this one, I can update the new helper function `findVPTInactiveOperandIdx` to use its new enum constant instead of having that ugly assert :slightly_smiling_face:

davemgreen commented on llvm/test/CodeGen/Thumb2/LowOverheadLoops/vctp-vs-unpredicated-copy.mir:
```diff
@@ -0,0 +1,231 @@
+
+--- |
+  ; ModuleID = '162644.c'
+  source_filename = "162644.c"
```
These tests can often be cleaned up a fair amount.  For example removing:
 - ModuleID and source_filename
 - Any `Function Attrs`
 - The intrinsic definitions
 - Sometimes the attributes and metadata
 - (Sometimes the entire llvm it partition can be emitted, but it is sometimes useful to keep to help explain the test.

davemgreen commented on llvm/test/CodeGen/Thumb2/LowOverheadLoops/vctp-vs-unpredicated-copy.mir:
```diff
@@ -0,0 +1,231 @@
+name:            test_func
+alignment:       4
+exposesReturnsTwice: false
+legalized:       false
```
Equally a lot of these can often be removed to help simplify the test case.

davemgreen reviewed (approved):
Thanks, this sounds sensible to me.

LGTM.

statham-arm committed:
Trim cruft from the new test case

davemgreen reviewed (approved):
Thanks, LGTM

statham-arm committed:
Merge from main to take advantage of #163273

statham-arm committed:
Use the new Tablegen sub-operand index constants

Both documents, and futureproofs, the offset from i to the operand we
want, so the assertion with its comment isn't needed any more.

statham-arm commented:
(Darn it, made a typo in the PR number in one of those commit messages!)
